src/System.ComponentModel.Annotations/tests/CompareAttributeTests.cs

```diff
@@ -2,100 +2,98 @@
-        {
-            var currentObject = new CompareObject("a");
-            var otherObject = new CompareObject("b");
+            yield return new object[] { nameof(CompareObject.CompareProperty), context, nameof(CompareObject.CompareProperty), typeof(ValidationException) };
+            yield return new object[] { nameof(CompareObject.ComparePropertyWithDisplayName), context, "DisplayName", typeof(ValidationException) };
+            yield return new object[] { "UnknownPropertyName", context, null, typeof(ValidationException) };
 
-            var testContext = new ValidationContext(otherObject, null, null);
-            testContext.DisplayName = "CurrentProperty";
+            ValidationContext subClassContext = new ValidationContext(new CompareObjectSubClass("a"));
+            yield return new object[] { nameof(CompareObject.CompareProperty), subClassContext, "CompareProperty", typeof(ValidationException) };
 
-            var attribute = new CompareAttribute("ComparePropertyWithDisplayName");
-            Assert.Throws<ValidationException>(
-                () => attribute.Validate(currentObject.CompareProperty, testContext));
+            yield return new object[] { "Item", context, null, typeof(TargetParameterCountException) };
+            yield return new object[] { nameof(CompareObject.SetOnlyProperty), context, null, typeof(ArgumentException) };
         }
 
-        [Fact]
-        public static void Validate_throws_when_PropertyName_is_unknown()
+        [Theory]
+        [MemberData(nameof(Invalid_TestData))]
+        public static void Validate_Invalid_Throws(string otherProperty, ValidationContext context, string otherPropertyDisplayName, Type exceptionType)
         {
-            var currentObject = new CompareObject("a");
-            var otherObject = new CompareObject("b");
+            var attribute = new CompareAttribute(otherProperty);
 
-            var testContext = new ValidationContext(otherObject, null, null);
-            testContext.DisplayName = "CurrentProperty";
+            string previousErrorMessage = attribute.FormatErrorMessage("name");
+            Assert.Throws(exceptionType, () => attribute.Validate("b", context));
+            Assert.Equal(otherPropertyDisplayName, attribute.OtherPropertyDisplayName);
 
-            var attribute = new CompareAttribute("UnknownPropertyName");
-            Assert.Throws<ValidationException>(
-                () => attribute.Validate(currentObject.CompareProperty, testContext));
-            // cannot check error message - not defined on ret builds
+            string newErrorMessage = attribute.FormatErrorMessage("name");
+            if (otherPropertyDisplayName == null || otherProperty.Equals(otherPropertyDisplayName))
+            {
+                Assert.Equal(previousErrorMessage, newErrorMessage);
+            }
+            else
+            {
+                Assert.NotEqual(previousErrorMessage, newErrorMessage);
+            }
```

Comment:
I'm a bit confused by this test.  Can you explain what it's doing?  Maybe split it up into different tests?  Apparently something is causing the result of FormatErrorMessage to potentially change, but it's not clear what (calling Validate?  Accessing OtherPropertyDisplayName?).  And none of that seems to jive with the name of the test, Validate_Invalid_Throws.